src/python_testing/TC_XX_1_1.py

```diff
@@ -0,0 +1,180 @@
+        wildcard_read = (await dev_ctrl.Read(self.dut_node_id, [()]))
+        endpoints_tlv = wildcard_read.tlvAttributes
+
+        self.print_step(2, "Verify check of invalid elements on all clusters, across all endpoints")
+        for endpoint_id, endpoint in endpoints_tlv.items():
+            logging.info(f"Verifying Endpoint: {endpoint_id}")
+
+            for cluster_id, cluster in endpoint.items():
+                logging.info(f"Verifying ClusterID: {cluster_id:#06x}")
+                logging.info(f"AttributeList: {cluster[ATTRIBUTE_LIST_ID]}")
+                logging.info(f"AcceptedCommandList: {cluster[ACCEPTED_COMMAND_LIST_ID]}")
+                logging.info(f"GeneratedCommandList: {cluster[GENERATED_COMMAND_LIST_ID]}")
+
+                attribute_list = cluster[ATTRIBUTE_LIST_ID]
+                # event_list = cluster[EVENT_LIST_ID]
+                accepted_command_list = cluster[ACCEPTED_COMMAND_LIST_ID]
+                generated_command_list = cluster[GENERATED_COMMAND_LIST_ID]
+
+                attribute_list.append(0x1000_4F00)
+
+                # Check for invalid elements in standard range
+                # -- Attributes
+                # print(f"Attribute invalid range: {(standard_range_prefix + attribute_id_range_max + 1):#010x} - {(standard_range_prefix + attribute_global_range_min):#010x} and {(standard_range_prefix + attribute_id_invalid_max):#010x}")
+
+                if any(((attribute_id > (standard_range_prefix + attribute_id_range_max) and attribute_id < (standard_range_prefix + attribute_global_range_min)) or attribute_id == attribute_id_invalid_max) for attribute_id in attribute_list):
+                    asserts.fail(f'Invalid attribute in range [{(standard_range_prefix + attribute_id_range_max + 1):#010x} - {(standard_range_prefix + attribute_global_range_min):#010x}] or [{attribute_id_invalid_max:#010x}]')
+
+                '''
+                # -- Events
+                # Currently disabled, since event list is not supported
+                # print(f"EventList invalid range: {(standard_range_prefix + event_id_range_max + 1):#010x} - {(standard_range_prefix + event_id_invalid_max):#010x}")
+
+                if any((event_id > (standard_range_prefix + event_id_range_max) and event_id <= (standard_range_prefix + event_id_invalid_max)) for event_id in event_list):
+                    asserts.fail(f'Invalid event in range [{(standard_range_prefix + event_id_range_max + 1):#010x} - {(standard_range_prefix + event_id_invalid_max):#010x}]')
+                '''
+
+                # -- Commands
+                # print(f"CommandLists invalid range: {(standard_range_prefix + command_id_range_max + 1):#010x} - {(standard_range_prefix + command_id_invalid_max):#010x}")
+
+                if any((accepted_command_id > (standard_range_prefix + command_id_range_max) and accepted_command_id <= (standard_range_prefix + command_id_invalid_max)) for accepted_command_id in accepted_command_list):
```

Comment:
This is very hard to parse.

I think this would be much clearer if we had:

1. Functions to extract the prefix/suffix.
2. Functions to check for suffix being in various valid ranges.
3. Functions to check for prefix being in various valid ranges.

Similar for the attributes cases above....  And given that we have to repeat everything for accepted/generated commands, ideally some sort of "valid command id" function that we can use for either....

Reply:
It is quite dense, I agree. 😄 

I will give it another spin and try to make it easier to read what it is actually doing as well as adding the suggested functions.

Reply:
Thanks for the great feedback, it made me rethink the approach and I think it got way simpler and easier to understand.